compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2352,19 +2353,30 @@ fn add_upstream_native_libraries(
                 NativeLibKind::Framework { as_needed } => {
                     cmd.link_framework(name, as_needed.unwrap_or(true))
                 }
-                NativeLibKind::Static { bundle: Some(false), .. } => {
+                NativeLibKind::Static { bundle: Some(false), whole_archive } => {
                     // Link "static-nobundle" native libs only if the crate they originate from
                     // is being linked statically to the current crate.  If it's linked dynamically
                     // or is an rlib already included via some other dylib crate, the symbols from
                     // native libs will have already been included in that dylib.
                     if data[cnum.as_usize() - 1] == Linkage::Static {
-                        cmd.link_staticlib(name, verbatim)
+                        if whole_archive == Some(true) {
+                            cmd.link_whole_staticlib(name, verbatim, &search_path);
+                        } else {
+                            cmd.link_staticlib(name, verbatim)
+                        }
                     }
                 }
                 // ignore statically included native libraries here as we've
                 // already included them when we included the rust library
                 // previously
-                NativeLibKind::Static { bundle: None | Some(true), .. } => {}
+                NativeLibKind::Static { bundle: None | Some(true), whole_archive } => {
+                    if whole_archive == Some(true) {
+                        bug!(
+                            "Combining `+bundle` with `+whole-archive` is not allowed. \
+                              This should have been caught by an earlier validation step already."
```

Comment:
I don't think there is such a validation step right now.
Is there a test for this case?

Reply:
I think @wesleywiser will be looking into this separately. Right, @wesleywiser?

Reply:
Yes, I'm working on this.

Reply:
@petrochenkov, does that resolve the issue for you or should we block this PR on the validation code landing?

Reply:
It seems questionable to knowingly introduce an ICE.

Reply:
OK, let's wait for @wesleywiser's changes to add the validation.

Reply:
I took the liberty of pushing into your PR @michaelwoerister.